File: frontend/src/services/borrowerService.ts
/**
 * Borrower API Service
 */

import api from './api';
import { logger } from '@/utils/logger';
import type {
  ApiResponse,
  PagedResponse,
  Borrower,
  CreateBorrowerRequest,
} from '@/types';

const borrowerLogger = logger.scope('BorrowerService');

export interface BorrowerFilters {
  page?: number;
  size?: number;
  sortBy?: string;
  sortDir?: 'asc' | 'desc';
  status?: string;
  search?: string;
}

export const borrowerService = {
  /**
   * Get all borrowers with pagination
   */
  async getAll(filters: BorrowerFilters = {}): Promise<PagedResponse<Borrower>['data']> {
    borrowerLogger.info('Fetching borrowers', { filters });

    const params = {
      page: filters.page ?? 0,
      size: filters.size ?? 20,
      sortBy: filters.sortBy ?? 'createdAt',
      sortDir: filters.sortDir ?? 'desc',
    };

    const response = await api.get<PagedResponse<Borrower>>('/v1/borrowers', { params });

    borrowerLogger.info('Borrowers fetched', {
      total: response.data.data.totalElements,
      page: response.data.data.page,
    });

    return response.data.data;
  },

  /**
   * Get borrower by ID
   */
  async getById(id: string): Promise<Borrower> {
    borrowerLogger.info('Fetching borrower by ID', { id });

    const response = await api.get<ApiResponse<Borrower>>(`/v1/borrowers/${id}`);

    borrowerLogger.info('Borrower fetched', {
      id: response.data.data.id,
      name: response.data.data.fullName,
    });

    return response.data.data;
  },

  /**
   * Get borrower by code
   */
  async getByCode(code: string): Promise<Borrower> {
    borrowerLogger.info('Fetching borrower by code', { code });

    const response = await api.get<ApiResponse<Borrower>>(`/v1/borrowers/code/${code}`);

    return response.data.data;
  },

  /**
   * Create new borrower
   */
  async create(data: CreateBorrowerRequest): Promise<Borrower> {
    borrowerLogger.info('Creating borrower', { name: data.fullName, phone: data.phone });

    const response = await api.post<ApiResponse<Borrower>>('/v1/borrowers', data);

    borrowerLogger.info('Borrower created', {
      id: response.data.data.id,
      code: response.data.data.borrowerCode,
    });

    return response.data.data;
  },

  /**
   * Update borrower
   */
  async update(id: string, data: Partial<CreateBorrowerRequest>): Promise<Borrower> {
    borrowerLogger.info('Updating borrower', { id, data });

    const response = await api.put<ApiResponse<Borrower>>(`/v1/borrowers/${id}`, data);

    borrowerLogger.info('Borrower updated', { id });

    return response.data.data;
  },

  /**
   * Block borrower
   */
  async block(id: string): Promise<Borrower> {
    borrowerLogger.info('Blocking borrower', { id });

    const response = await api.post<ApiResponse<Borrower>>(`/v1/borrowers/${id}/block`);

    borrowerLogger.info('Borrower blocked', { id });

    return response.data.data;
  },

  /**
   * Activate borrower
   */
  async activate(id: string): Promise<Borrower> {
    borrowerLogger.info('Activating borrower', { id });

    const response = await api.post<ApiResponse<Borrower>>(`/v1/borrowers/${id}/activate`);

    borrowerLogger.info('Borrower activated', { id });

    return response.data.data;
  },

  /**
   * Delete borrower (soft delete)
   */
  async delete(id: string): Promise<void> {
    borrowerLogger.info('Deleting borrower', { id });

    await api.delete(`/v1/borrowers/${id}`);

    borrowerLogger.info('Borrower deleted', { id });
  },

  /**
   * Search borrowers
   */
  async search(query: string, page = 0, size = 20): Promise<PagedResponse<Borrower>['data']> {
    borrowerLogger.info('Searching borrowers', { query, page, size });

    const response = await api.get<PagedResponse<Borrower>>('/v1/borrowers/search', {
      params: { q: query, page, size },
    });

    borrowerLogger.info('Search results', {
      query,
      total: response.data.data.totalElements,
    });

    return response.data.data;
  },

  /**
   * Get borrowers by status
   */
  async getByStatus(
    status: string,
    page = 0,
    size = 20
  ): Promise<PagedResponse<Borrower>['data']> {
    borrowerLogger.info('Fetching borrowers by status', { status, page, size });

    const response = await api.get<PagedResponse<Borrower>>(
      `/v1/borrowers/status/${status}`,
      { params: { page, size } }
    );

    return response.data.data;
  },
};

export default borrowerService;

File: loan-platform/src/main/java/com/loanplatform/borrower/service/BorrowerService.java
package com.loanplatform.borrower.service;

import com.loanplatform.borrower.dto.BorrowerResponse;
import com.loanplatform.borrower.dto.CreateBorrowerRequest;
import com.loanplatform.borrower.dto.UpdateBorrowerRequest;
import com.loanplatform.borrower.entity.Borrower;
import com.loanplatform.borrower.entity.BorrowerStatus;
import com.loanplatform.borrower.entity.RiskBand;
import com.loanplatform.borrower.mapper.BorrowerMapper;
import com.loanplatform.borrower.repository.BorrowerRepository;
import com.loanplatform.common.config.TenantContext;
import com.loanplatform.common.dto.PagedResponse;
import com.loanplatform.common.exception.BusinessException;
import com.loanplatform.common.exception.ResourceNotFoundException;
import com.loanplatform.tenant.repository.TenantRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;
import java.util.concurrent.atomic.AtomicLong;

@Service
@RequiredArgsConstructor
@Slf4j
public class BorrowerService {

    private final BorrowerRepository borrowerRepository;
    private final BorrowerMapper borrowerMapper;
    private final TenantRepository tenantRepository;

    private final AtomicLong borrowerCodeSequence = new AtomicLong(System.currentTimeMillis() % 100000);

    @Transactional
    public BorrowerResponse createBorrower(CreateBorrowerRequest request, UUID createdByUserId) {
        UUID tenantId = TenantContext.requireTenant();
        log.info("Creating borrower for tenant: {}", tenantId);

        validateTenantLimits(tenantId);

        if (borrowerRepository.existsByPhoneAndTenantIdAndDeletedFalse(request.getPhone(), tenantId)) {
            throw new BusinessException("Borrower with this phone already exists", "PHONE_EXISTS");
        }

        if (request.getIdNumber() != null &&
            borrowerRepository.existsByIdNumberAndTenantIdAndDeletedFalse(request.getIdNumber(), tenantId)) {
            throw new BusinessException("Borrower with this ID number already exists", "ID_NUMBER_EXISTS");
        }

        Borrower borrower = borrowerMapper.toEntity(request);
        borrower.setTenantId(tenantId);
        borrower.setBorrowerCode(generateBorrowerCode(tenantId));
        borrower.setCreatedBy(createdByUserId);

        calculateRiskScore(borrower);

        borrower = borrowerRepository.save(borrower);
        log.info("Borrower created: {} for tenant: {}", borrower.getId(), tenantId);

        return borrowerMapper.toResponse(borrower);
    }

    @Transactional(readOnly = true)
    public BorrowerResponse getBorrowerById(UUID borrowerId) {
        UUID tenantId = TenantContext.requireTenant();
        Borrower borrower = borrowerRepository.findByIdAndTenantId(borrowerId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Borrower", borrowerId));
        return borrowerMapper.toResponse(borrower);
    }

    @Transactional(readOnly = true)
    public BorrowerResponse getBorrowerByCode(String borrowerCode) {
        UUID tenantId = TenantContext.requireTenant();
        Borrower borrower = borrowerRepository.findByBorrowerCodeAndTenantId(borrowerCode, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Borrower", borrowerCode));
        return borrowerMapper.toResponse(borrower);
    }

    @Transactional(readOnly = true)
    public PagedResponse<BorrowerResponse> getAllBorrowers(int page, int size, String sortBy, String sortDir) {
        UUID tenantId = TenantContext.requireTenant();

        Sort sort = sortDir.equalsIgnoreCase("desc")
                ? Sort.by(sortBy).descending()
                : Sort.by(sortBy).ascending();

        Pageable pageable = PageRequest.of(page, size, sort);
        Page<Borrower> borrowers = borrowerRepository.findAllByTenantId(tenantId, pageable);

        return PagedResponse.of(
                borrowers.getContent().stream()
                        .map(borrowerMapper::toResponse)
                        .toList(),
                page,
                size,
                borrowers.getTotalElements()
        );
    }

    @Transactional(readOnly = true)
    public PagedResponse<BorrowerResponse> searchBorrowers(String search, int page, int size) {
        UUID tenantId = TenantContext.requireTenant();
        Pageable pageable = PageRequest.of(page, size, Sort.by("fullName").ascending());
        Page<Borrower> borrowers = borrowerRepository.searchBorrowers(tenantId, search, pageable);

        return PagedResponse.of(
                borrowers.getContent().stream()
                        .map(borrowerMapper::toResponse)
                        .toList(),
                page,
                size,
                borrowers.getTotalElements()
        );
    }

    @Transactional(readOnly = true)
    public PagedResponse<BorrowerResponse> getBorrowersByStatus(BorrowerStatus status, int page, int size) {
        UUID tenantId = TenantContext.requireTenant();
        Pageable pageable = PageRequest.of(page, size, Sort.by("fullName").ascending());
        Page<Borrower> borrowers = borrowerRepository.findByTenantIdAndStatus(tenantId, status, pageable);

        return PagedResponse.of(
                borrowers.getContent().stream()
                        .map(borrowerMapper::toResponse)
                        .toList(),
                page,
                size,
                borrowers.getTotalElements()
        );
    }

    @Transactional
    public BorrowerResponse updateBorrower(UUID borrowerId, UpdateBorrowerRequest request, UUID updatedByUserId) {
        UUID tenantId = TenantContext.requireTenant();
        Borrower borrower = borrowerRepository.findByIdAndTenantId(borrowerId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Borrower", borrowerId));

        if (request.getPhone() != null && !request.getPhone().equals(borrower.getPhone())) {
            if (borrowerRepository.existsByPhoneAndTenantIdAndDeletedFalse(request.getPhone(), tenantId)) {
                throw new BusinessException("Borrower with this phone already exists", "PHONE_EXISTS");
            }
        }

        borrowerMapper.updateEntity(borrower, request);
        borrower.setUpdatedBy(updatedByUserId);

        calculateRiskScore(borrower);

        borrower = borrowerRepository.save(borrower);
        log.info("Borrower updated: {}", borrowerId);

        return borrowerMapper.toResponse(borrower);
    }

    @Transactional
    public void blockBorrower(UUID borrowerId, UUID blockedByUserId) {
        UUID tenantId = TenantContext.requireTenant();
        Borrower borrower = borrowerRepository.findByIdAndTenantId(borrowerId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Borrower", borrowerId));

        borrower.block();
        borrower.setUpdatedBy(blockedByUserId);
        borrowerRepository.save(borrower);
        log.info("Borrower blocked: {}", borrowerId);
    }

    @Transactional
    public void activateBorrower(UUID borrowerId, UUID activatedByUserId) {
        UUID tenantId = TenantContext.requireTenant();
        Borrower borrower = borrowerRepository.findByIdAndTenantId(borrowerId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Borrower", borrowerId));

        borrower.activate();
        borrower.setUpdatedBy(activatedByUserId);
        borrowerRepository.save(borrower);
        log.info("Borrower activated: {}", borrowerId);
    }

    @Transactional
    public void deleteBorrower(UUID borrowerId, UUID deletedByUserId) {
        UUID tenantId = TenantContext.requireTenant();
        Borrower borrower = borrowerRepository.findByIdAndTenantId(borrowerId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Borrower", borrowerId));

        borrower.softDelete(deletedByUserId);
        borrowerRepository.save(borrower);
        log.info("Borrower soft-deleted: {}", borrowerId);
    }

    public long countBorrowersByTenant(UUID tenantId) {
        return borrowerRepository.countByTenantId(tenantId);
    }

    private String generateBorrowerCode(UUID tenantId) {
        String prefix = tenantId.toString().substring(0, 4).toUpperCase();
        long seq = borrowerCodeSequence.incrementAndGet();
        return String.format("BRW-%s-%06d", prefix, seq % 1000000);
    }

    private void validateTenantLimits(UUID tenantId) {
        var tenant = tenantRepository.findById(tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Tenant", tenantId));

        long currentCount = borrowerRepository.countByTenantId(tenantId);
        if (currentCount >= tenant.getMaxBorrowers()) {
            throw new BusinessException("Maximum borrower limit reached for this tenant", "LIMIT_EXCEEDED");
        }
    }

    private void calculateRiskScore(Borrower borrower) {
        int score = 50;

        if (borrower.getMonthlyIncome() != null) {
            if (borrower.getMonthlyIncome().doubleValue() > 50000) {
                score += 20;
            } else if (borrower.getMonthlyIncome().doubleValue() > 25000) {
                score += 10;
            }
        }

        if (borrower.getIdNumber() != null && !borrower.getIdNumber().isBlank()) {
            score += 10;
        }

        if (borrower.getEmail() != null && !borrower.getEmail().isBlank()) {
            score += 5;
        }

        if (borrower.getEmployerName() != null && !borrower.getEmployerName().isBlank()) {
            score += 5;
        }

        borrower.setRiskScore(score);

        if (score >= 80) {
            borrower.setRiskBand(RiskBand.LOW);
        } else if (score >= 60) {
            borrower.setRiskBand(RiskBand.MEDIUM);
        } else if (score >= 40) {
            borrower.setRiskBand(RiskBand.HIGH);
        } else {
            borrower.setRiskBand(RiskBand.VERY_HIGH);
        }
    }
}
